internal/rbd/types/mirror.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+import (
+	"context"
+	"time"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/ceph/go-ceph/rbd/admin"
+)
+
+// FlattenMode is used to indicate the flatten mode for an RBD image.
+type FlattenMode string
+
+const (
+	// FlattenModeNever indicates that the image should never be flattened.
+	FlattenModeNever FlattenMode = "never"
+	// FlattenModeForce indicates that the image with the parent must be flattened.
+	FlattenModeForce FlattenMode = "force"
+)
+
+type Mirror interface {
+	// EnableMirroring enables mirroring on the resource with the specified mode.
+	EnableMirroring(mode librbd.ImageMirrorMode) error
+	// DisableMirroring disables mirroring on the resource with the option to force the operation
+	DisableMirroring(force bool) error
+	// Promote promotes the resource to primary status with the option to force the operation
+	Promote(force bool) error
+	// ForcePromote promotes the resource to primary status with a timeout
+	ForcePromote(cr *util.Credentials) error
+	// Demote demotes the resource to secondary status
+	Demote() error
+	// Resync resynchronizes the resource
+	Resync() error
+	// GetMirroringInfo returns the mirroring information of the resource
+	GetMirroringInfo() (MirrorInfo, error)
+	// GetMirroringInfo returns the mirroring information of the resource
+	GetGlobalMirroringStatus() (GlobalStatus, error)
+	// AddSnapshotScheduling adds a snapshot scheduling to the resource
+	AddSnapshotScheduling(interval admin.Interval, startTime admin.StartTime) error
+}
+
+type MirrorInfo interface {
+	// IsPrimary returns true if the resource is primary
+	IsPrimary() bool
+	// GetState returns the state of the resource
+	GetState() string
```

Comment:
consider adding a `MirrorState` type to replace the string value

Reply:
This needs more work, will take it in next update when implementing this for group